### PR TITLE
Add bindings for iconv calls

### PR DESCRIPTION
**Pull request by Minoru**

### CODE REVIEW

**Comment by Minoru** [comment]
Already looking into the macOS failure!

**Comment by Minoru** [comment]
Sorry, I need a bit of help with this ­— I either misunderstand `cfg_if!` macro or `target_os` attribute.

The original commit, https://github.com/rust-lang/libc/pull/2037/commits/ac6ec1c88e0b3a439703777508cd254bbacb818c, failed to build on macOS because it didn't link with `iconv` library. The next commit, added an appropriate `link` attribute, but the build still failed — **and I don't see `-liconv` anywhere in the output**. Adding a `link` attribute unconditionally works on macOS, but not other platforms (https://github.com/rust-lang/libc/pull/2037/commits/ddddc18f1d8ba23c133e5310052e9be2aaf5afeb), the latter being expected.

macOS build links with `m` and `c` libraries, and the only other place where those could come from is the "Android or OpenBSD" branch. But if I add iconv there, still no `-liconv` in the output (https://github.com/rust-lang/libc/pull/2037/commits/8d7a53c6085e8d8c57f48b8becdff1550d64a23f).

It feels like I'm adding the `link` attribute to the wrong place. Can anyone point me to the right one?

**Comment by Minoru** [comment]
https://github.com/rust-lang/libc/pull/2037/commits/99202f737ca56714c6ffe0d687c945feb9b185ed implements a new approach: build.rs emits `rustc-link-lib=iconv` on Apple platforms. This works, but might be confusing to future programmers, since all the other libraries are configured via the `link` attributes.

File: src/unix/mod.rs [hunk 1]
```diff
@@ -298,6 +299,7 @@ cfg_if! {
     } else if #[cfg(feature = "std")] {
         // cargo build, don't pull in anything extra as the libstd dep
         // already pulls in all libs.
+        extern {}
```
Comment (joshtriplett) [inline src/unix/mod.rs:302]: Could you clarify why you've added this?

Comment (Minoru) [reply src/unix/mod.rs:302]: Ah, sorry, it's a leftover from one of the experiments, https://github.com/rust-lang/libc/pull/2037/commits/355301417300e33fd48571f4c0248c7f5db9254c. Will remove this in a second.

**Comment by Minoru** [comment]
Even though this PR passes CI, I'd really like some discussion on whether this even belongs to libc.

On one hand it definitely does, since iconv calls are part of POSIX, and are widely implemented.

On the other hand, the situation is a bit complicated since BDSs can have *two* implementations at once: one from libc.so (implementing POSIX), the other from libiconv.so (GNU libiconv implementing POSIX plus GNU extensions). libiconv uses different symbol names (`libiconv_open()` etc.), but the function names are the same, so the caller is none the wiser — the decision is made at linking time. (musl might not implement the GNU extensions either, I didn't check.)

I doubt it's the crate's job to decide which iconv implementation my app should link to, and so there are a few possible solutions:
1. don't provide these bindings at all, let some other crate do it;
2. provide a feature flag to pick between "libc iconv" and "libiconv iconv", *regardless of the platform*;
3. provide a feature flag to pick between "libc iconv" and "libiconv iconv", *just on systems that don't use glibc* (I believe the libc "type" is a part of a platform triple);
4. link to "libc iconv" unconditionally, let someone else provide libiconv bindings;
5. link to "libiconv iconv" unconditionally, let someone else provide libc.so bindings.

I only discovered all these complications once I opened the PR; if I knew of them in advance, I'd open an issue first. Sorry about that! Happy to close this and re-report as an issue if that's what you'd prefer.

**Comment by semarie** [comment]
at least on OpenBSD (but I think it is true for several others OS), iconv isn't part of system libc (neither part of libraries provided by the system). iconv is only available as third party library.

**Comment by Minoru** [comment]
I still want to have the discussion, so I'll hold off making any changes until it's decided if these bindings have a place in this crate.

**Comment by JohnTitor** [comment]
Thanks for clarifying, hmm, then I doubt we should this to the UNIX module. And the tweak for Apple targets is tricky, as you said, so it should have some comments at least, I believe.

/cc @Amanieu Any thoughts on this?

**Comment by Amanieu** [comment]
I'm thinking option 4 might be the best:

> link to "libc iconv" unconditionally, let someone else provide libiconv bindings;

However we should only provide bindings if this API is actually provided by libc. So for example this wouldn't be present on OpenBSD.

**Comment by Minoru** [comment]
All done, please take a look.

File: libc-test/build.rs [hunk 2]
```diff
@@ -360,6 +361,7 @@ fn test_openbsd(target: &str) {
         "pthread_np.h",
         "sys/syscall.h",
         "sys/shm.h",
+        "iconv.h",
```
Comment (semarie) [inline libc-test/build.rs:364]: this break the test on OpenBSD : no `iconv.h` file here

Comment (semarie) [reply libc-test/build.rs:364]: I just made #2067 to unbreak OpenBSD

Comment (Minoru) [reply libc-test/build.rs:364]: Oops, sorry — that's leftovers from the time when the bindings were in the unix mod. The similar problem affects Android, Solari-sh, Redox and a few other targets, I'll submit a PR fixing that.

Comment (Minoru) [reply libc-test/build.rs:364]: Done: https://github.com/rust-lang/libc/pull/2068

**Pull request merged**